[tests] Refactor SqlServer tests to use testcontainers

radical opened the pull request:
Also add tracing tests for `Aspire.Microsoft.Data.SqlClient.Tests`

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/3034)

radical committed:
[tests] Refactor SqlServer tests to use testcontainers

radical committed:
Add SqlServerContainerImageTags

radical commented:
@eerhardt Are there any tests that I missed here?

radical committed:
Aspire.Microsoft.Data.SqlClient.Tests: add tracing tests, and use testcontainers

radical committed:
Run sqlclient tests on helix

eerhardt commented on src/Aspire.Hosting.SqlServer/Aspire.Hosting.SqlServer.csproj:
```diff
@@ -14,6 +14,8 @@
   <ItemGroup>
     <ProjectReference Include="..\Aspire.Hosting\Aspire.Hosting.csproj" />
+
+    <InternalsVisibleTo Include="Aspire.Hosting.Tests" />
```
Can we just compile the file into the tests instead? I know other `Aspire.Hosting.X` assemblies do this, but they need it for other reasons and I'd like to limit when we use IVT.

eerhardt commented on tests/Aspire.Microsoft.EntityFrameworkCore.SqlServer.Tests/SqlServerContainerFixture.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Aspire.SqlServer.Tests;
+
+public sealed class SqlServerContainerFixture : IAsyncLifetime
```
(nit) The PostgreSQLContainerFixture goes in the non-EF tests and then the EF tests links to it. Can we follow that same pattern here? Have the EF thing depend on the non-EF thing seems like the right layering.

radical replied:
Would that be `Aspire.Microsoft.Data.SqlClient.Tests`?

eerhardt commented on tests/Aspire.Microsoft.Data.SqlClient.Tests/ConformanceTests.cs:
```diff
@@ -1,23 +1,35 @@
-    // TODO
-    protected override string[] RequiredLogCategories => Array.Empty<string>();
+    protected override string[] RequiredLogCategories => [
+        "Microsoft.Extensions.Hosting.Internal.ApplicationLifetime",
```
Were tests failing if we don't do this? We don't have these categories elsewhere.

The issue is that Microsoft.Data.SqlClient doesn't support logging. See `Aspire.Microsoft.Data.SqlClient:` in https://github.com/dotnet/aspire/blob/main/src/Components/Telemetry.md.

radical replied:
No, the tests weren't failing because they didn't check for any required categories.

From your link:
```
Aspire.Microsoft.EntityFrameworkCore.SqlServer:

Log categories:
"Microsoft.EntityFrameworkCore.ChangeTracking"
"Microsoft.EntityFrameworkCore.Database.Command"
"Microsoft.EntityFrameworkCore.Database.Connection"
"Microsoft.EntityFrameworkCore.Database.Transaction"
"Microsoft.EntityFrameworkCore.Infrastructure"
"Microsoft.EntityFrameworkCore.Migrations"
"Microsoft.EntityFrameworkCore.Model"
"Microsoft.EntityFrameworkCore.Model.Validation"
"Microsoft.EntityFrameworkCore.Query"
"Microsoft.EntityFrameworkCore.Update"
```

Shouldn't we check for these?

radical replied:
oops, you were talking about `SqlClient`:
```
Aspire.Microsoft.Data.SqlClient:
Log categories:
none (the client does not provide an easy way to integrate it with logger factory)
```

radical committed:
Merge remote-tracking branch 'origin/main' into sqlserver-testcontainers

radical committed:
Address review feedback from @ eerhardt

- move SqlServerContainerFixture to `Aspire.Microsoft.Data.SqlClient.Tests`
- Instead of using `InternalsVisibleTo`, compile the relevant file with
  the project.

radical committed:
Update SqlServer/SqlClient ActivitySourceName from OpenTelemetry.Instrumentation.SqlClient to Aspire.Microsoft.EntityFrameworkCore.SqlServer

radical committed:
Remove RequiredLogCategories for SqlClient as none are supported, see src/Components/Telemetry.md

radical committed:
Update Activity source names for SqlClient, and SqlServer in Telemetry.md

radical commented:
I made the suggested changes. And also had to update `ActivitySourceName`:
```diff
-    protected override string ActivitySourceName => "OpenTelemetry.Instrumentation.SqlClient";
+    protected override string ActivitySourceName => "Aspire.Microsoft.Data.SqlClient";
```
```diff
-    protected override string ActivitySourceName => "OpenTelemetry.Instrumentation.SqlClient";
+    protected override string ActivitySourceName => "Aspire.Microsoft.EntityFrameworkCore.SqlServer";
```

does that change look correct?

radical committed:
Merge branch 'main' into sqlserver-testcontainers

normj committed:
Fix typo in method name "Forwaded" -> "Forwarded" (#3123)

davidfowl committed:
Change how resources expose urls to the dashboard (#3103)

* Change how resources expose urls to the dashboard
- The resource server used to have a services and
endpoints split, that is now a single object called a url
which has a name, a full url (uri format) and a bool for
if that endpoint is internal (or an implementation detail).
- Clean up the way we extract urls from the dcp information by
mapping the app model resource and endpoint within each service
resource in DCP.
- Mark fields as deprecated in the proto file
- Moved from Tuples to records for CustomResourceSnapshot

timheuer committed:
Set the value to be the Text on Endpoints (#3130)

Fix #3128

drewnoakes committed:
Auth for dashboard web app and resource service client (#3033)

* Enable basic certificate auth for resource service

This adds the ability to configured the gRPC connection from the dashboard to a resource service to use certificates for authentication.

Such auth is not used in the local dev scenario, so the app host is changed to suppress auth via the `DOTNET_RESOURCE_SERVICE_DISABLE_AUTH` environment variable. The dashboard will now require certificates unless this variable is set to `1`. This discourages self-hosting the dashboard in an insecure manner.

* Support loading client cert from keystore

* Use enum for resource service auth mode

* Dashboard web app supports OpenID Connect

* Use a single authentication call for both the web app and OTLP

Use explicit policies and configure those policies with the authentication schemes required.

Also rejig some config settings for consistency.

* Remove redundant configuration key

* Rename "web app" to "frontend"

* Reorder readme sections and adjust headings and their levels

* Seal class

* Simplify frontend auth implementation

* Add docs

* More thorough enum validation

* Remove unused config key

* Change exception type

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/arcade build (#3140)

[main] Update dependencies from dotnet/arcade

mitchdenny committed:
Fix PrincipalId/PrincipalType excess parameters. (#3135)

dotnet-maestro[bot] committed:
Update dependencies from https://github.com/dotnet/arcade build (#3148)

[main] Update dependencies from dotnet/arcade

drewnoakes committed:
Fix key store config binding (#3139)

This was missed in #3033.

eerhardt commented on tests/Aspire.Microsoft.Data.SqlClient.Tests/ConformanceTests.cs:
```diff
@@ -1,22 +1,29 @@
 
     // https://github.com/open-telemetry/opentelemetry-dotnet/blob/031ed48714e16ba4a5b099b6e14647994a0b9c1b/src/OpenTelemetry.Instrumentation.SqlClient/Implementation/SqlActivitySourceHelper.cs#L31
-    protected override string ActivitySourceName => "OpenTelemetry.Instrumentation.SqlClient";
+    protected override string ActivitySourceName => "Aspire.Microsoft.Data.SqlClient";
```
This shouldn't be updated. It needs https://github.com/dotnet/aspire/pull/3118 which will fix this.

eerhardt commented on src/Components/Telemetry.md:
```diff
@@ -95,7 +95,7 @@ Aspire.Microsoft.Data.SqlClient:
   - none (the client does not provide an easy way to integrate it with logger factory)
 - Activity source names:
-  - "OpenTelemetry.Instrumentation.SqlClient"
+  - "Aspire.Microsoft.Data.SqlClient"
```
This should be reverted. https://github.com/dotnet/aspire/pull/3118 will fix it.

eerhardt commented on tests/Aspire.Microsoft.Data.SqlClient.Tests/ConformanceTests.cs:
```diff
@@ -71,5 +81,26 @@ protected override void SetMetrics(MicrosoftDataSqlClientSettings options, bool
         => options.Metrics = enabled;
 
     protected override void TriggerActivity(SqlConnection service)
-        => service.Open();
```
Was just opening a connection not enough to trigger an activity?

radical replied:
Yeah, nothing was emitted with just that, and the tests would fail:
```
[xUnit.net 00:00:23.52]     Aspire.Microsoft.Data.SqlClient.Tests.ConformanceTests.TracingEnablesTheRightActivitySource [FAIL]
  Failed Aspire.Microsoft.Data.SqlClient.Tests.ConformanceTests.TracingEnablesTheRightActivitySource [7 s]
  Error Message:
   Microsoft.DotNet.RemoteExecutor.RemoteExecutionException : Remote process failed with an unhandled exception.
  Stack Trace:

Child exception:
  Xunit.Sdk.NotEmptyException: Assert.NotEmpty() Failure
   at Aspire.Components.ConformanceTests.ConformanceTests`2.ActivitySourceTest(String key) in /Users/ankj/dev/a3/tests/Aspire.Components.Common.Tests/ConformanceTests.cs:line 405
   at Aspire.Microsoft.Data.SqlClient.Tests.ConformanceTests.<>c.<TracingEnablesTheRightActivitySource>b__24_1(ConformanceTests obj) in /Users/ankj/dev/a3/tests/Aspire.Microsoft.Data.SqlClient.Tests/ConformanceTests.cs:line 94
   at Aspire.Microsoft.Data.SqlClient.Tests.ConformanceTests.RunWithFixtureAsync(Action`1 test) in /Users/ankj/dev/a3/tests/Aspire.Microsoft.Data.SqlClient.Tests/ConformanceTests.cs:line 104
   at Aspire.Microsoft.Data.SqlClient.Tests.ConformanceTests.RunWithFixtureAsync(Action`1 test) in /Users/ankj/dev/a3/tests/Aspire.Microsoft.Data.SqlClient.Tests/ConformanceTests.cs:line 104
   at Microsoft.DotNet.RemoteExecutor.Program.Main(String[] args) in /_/src/Microsoft.DotNet.RemoteExecutor/src/Program.cs:line 65

Child process:
  Aspire.Microsoft.Data.SqlClient.Tests, Version=42.42.42.42, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51 Aspire.Microsoft.Data.SqlClient.Tests.ConformanceTests+<>c System.Threading.Tasks.Task <TracingEnablesTheRightActivitySource>b__24_0()
```

sebastienros committed:
Fix SqlClient component activity name (#3118)

* Fix SqlClient component activity name

* Update instructions

eerhardt committed:
Fix KeyVault README (#3151)

Follow up to #3120

radical committed:
address review feedback

radical committed:
Merge remote-tracking branch 'origin/main' into sqlserver-testcontainers

# Conflicts:
#	eng/Versions.props
#	src/Aspire.Hosting/Dcp/ApplicationExecutor.cs

radical committed:
Fix merge

eerhardt reviewed (approved):
LGTM. Thanks!
